Library/Homebrew/dev-cmd/linkage.rb

```diff
@@ -0,0 +1,127 @@
+#                breakage was found.
+
+require "set"
+require "keg"
+require "formula"
+
+module Homebrew
+  def linkage
+    found_broken_dylibs = false
+    ARGV.kegs.each do |keg|
+      ohai "Checking #{keg.name} linkage" if ARGV.kegs.size > 1
+      result = LinkageChecker.new(keg)
+      if ARGV.include?("--test")
+        result.display_test_output
+      else
+        result.display_normal_output
+      end
+      found_broken_dylibs = true unless result.broken_dylibs.empty?
+    end
+    if ARGV.include?("--test") && found_broken_dylibs
+      exit 1
+    end
+  end
+
+  class LinkageChecker
+    attr_reader :keg
+    attr_reader :broken_dylibs
+
+    def initialize(keg)
+      @keg = keg
+      @brewed_dylibs = Hash.new { |h, k| h[k] = Set.new }
+      @system_dylibs = Set.new
+      @broken_dylibs = Set.new
+      # dylib install paths using '@rpath' or other prefix variables
+      @variable_dylibs = Set.new
+      check_dylibs
+    end
+
+    def check_dylibs
+      @keg.find do |file|
+        next unless file.dylib? || file.mach_o_executable? || file.mach_o_bundle?
+        file.dynamically_linked_libraries.each do |dylib|
+          if !dylib.empty? && dylib[0] == "@"
+            @variable_dylibs << dylib
+          else
+            begin
+              owner = Keg.for Pathname.new(dylib)
+            rescue NotAKegError
+              @system_dylibs << dylib
+            rescue Errno::ENOENT
+              @broken_dylibs << dylib
+            else
+              @brewed_dylibs[owner.name] << dylib
+            end
+          end
+        end
+      end
+
+      begin
+        f = Formula[keg.name]
+        @undeclared_deps = @brewed_dylibs.keys - f.deps.map(&:name)
+        @undeclared_deps -= [f.name]
+      rescue FormulaUnavailableError
+        opoo "Formula unavailable: #{keg.name}"
+        @undeclared_deps = []
+      end
+    end
+
+    def display_normal_output
+      unless @system_dylibs.empty?
+        puts "System libraries:"
+        @system_dylibs.sort.each do |dylib|
+          puts "  #{dylib}"
+        end
+      end
+
+      unless @brewed_dylibs.empty?
+        puts "Homebrew libraries:"
+        @brewed_dylibs.sort.each do |name, dylibs|
+          dylibs.each do |dylib|
+            puts "  #{dylib} (#{name})"
+          end
+        end
+      end
+
+      unless @variable_dylibs.empty?
+        puts "Variable-referenced libraries:"
+        @variable_dylibs.sort.each do |dylib|
+          puts "  #{dylib}"
+        end
+      end
+
+      unless @broken_dylibs.empty?
+        puts "Missing libraries:"
+        @broken_dylibs.sort.each do |dylib|
+          puts "  #{dylib}"
+        end
+      end
+
+      unless @undeclared_deps.empty?
+        puts "Possible undeclared dependencies:"
+        puts "  #{@undeclared_deps * ", "}"
+      end
+    end
+
+    def display_test_output
+      if @broken_dylibs.empty?
+        puts "No broken dylib links"
+      else
+        puts "Broken dylib links:"
```

Comment:
I find it confusing that the terminology is different depending on whether `--test` was supplied or not. Here it's “Broken” but otherwise it's “Missing”. Could also be worthwhile to split `LinkageChecker#display_normal_output` into smaller methods and reuse the code here.